Filter route hints when creating invoices

ViktorTigerstrom opened the pull request:
Filter route hints when creating invoices as described in issue #1279 .

Just a few things worth mentioning:

For `create_phantom_invoice` I chose not to implement filtering if any of the channels are private, as from my understanding of the phantom node/invoice concept, that would make no sense. If my understanding is incorrect, please let me know.

The test code can be cleaned up a bit, by removing some recurring code. I wanted to push out the filtering code before taking care of that though, to find out if there's anything I need to improve with the filtering. Is there some macro to match channels with the invoice route hints, or would it make sense for me to make one?

Something worth mentioning is that as noted in the issue comment I made, this filtering logic when we filter out all channels that don't exceed or match the invoice amount, makes it easier for an external entity to use the invoice creation to potentially determine the inbound capacity of the channel. This can be seen in the `test_hints_has_no_channels_with_lower_inbound_capacity_than_invoice_amt` test. If this is not acceptable, I'd suggest that we'd remove this specific filtering and always return the highest capacity channel regardless of it's capacity. 

Other than that, I'd more than happy to address any feedback!

ViktorTigerstrom commented:
Oh sorry, some checks fails. I'll update asap

TheBlueMatt commented on lightning-invoice/src/utils.rs:
```diff
@@ -62,9 +62,17 @@ pub fn create_phantom_invoice<Signer: Sign, K: Deref>(
 	if let Some(amt) = amt_msat {
 		invoice = invoice.amount_milli_satoshis(amt);
 	}
+	let mut route_hints: HashMap<PublicKey, (u64, RouteHint)> = HashMap::new();
```
For phantoms, this should be per-"real"-node, not across the full outer loop (this will make it more similar to the non-phantom case as well).

ViktorTigerstrom replied:
Thanks!

TheBlueMatt reviewed (commented):
Hmm, it'd be really nice to DRY this up somewhat, either via some method that accepts a `&[ChannelDetails]` and returns an iterator over `ChannelDetails`'s to use, or a macro if you can't figure out the typing on that.

ViktorTigerstrom commented:
> Hmm, it'd be really nice to DRY this up somewhat, either via some method that accepts a `&[ChannelDetails]` and returns an iterator over `ChannelDetails`'s to use, or a macro if you can't figure out the typing on that.

Sure, I'll DRY that and the tests up!

ViktorTigerstrom commented:
Just pushed some DRY changes! Hope the trade-offs I had to make in regards to code readability are ok :)

TheBlueMatt commented on lightning-invoice/src/utils.rs:
```diff
@@ -218,6 +215,81 @@ where
+fn filter_channels<F1, F2>(
+	channels: Vec<ChannelDetails>,
+	min_inbound_capacity_required: Option<u64>,
+	empty_result_if_private_channels_exists: bool,
```
Hmm, I'm not sure if this captures what we want exactly, if our channels are public and we're creating a phantom route hint, instead of including all the channels, we can instead just create route hints that are one hop instead of two. id instead of creating the route hint in a callback, we can instead have the `create_phantom_invoice` callsite push a second hop onto all the route hints we get (or, if we get none, create a single route hint that is just one hop long). Then we could remove this option, I believe.

ViktorTigerstrom replied:
Thanks! If my new understanding explained in the logic below is correct, we can indeed remove it.

TheBlueMatt commented on lightning-invoice/src/utils.rs:
```diff
@@ -218,6 +215,81 @@ where
+	channels: Vec<ChannelDetails>,
+	min_inbound_capacity_required: Option<u64>,
+	empty_result_if_private_channels_exists: bool,
+	channel_pubkey_selector: F1,
```
Given this always returns a constant (ie cause we call `filter_channels` inside the per-node loop in `create_phantom_invoice`), do we need this at all?

ViktorTigerstrom replied:
In the `create_invoice_from_channelmanager` it actaully will vary, given that the channel counterparty node is different. However if my new understanding explained below is correct, it will always be the channel counterparty node id we are interested in, and this parameter wouldn't be needed at all! 
That would make the code a lot cleaner :)

TheBlueMatt reviewed (commented):
Cleanups look quite nice. A few comments that may enable further cleanups.

ViktorTigerstrom commented:
@TheBlueMatt Thanks a lot for the feedback!
Hmmmm, I've realized that I probably had an incorrect understanding of how the phantom invoice hints should be filtered. Sorry about that! Therefore I just to ensure that my current understanding is correct so I don't end up repeating my mistakes. Just to clarify, imagine the following scenario:

We have two nodes that are our phantom payment-receiving nodes that share same `cross_node_seed`:

- node1
- node2

Our nodes have channels connected to two counterparty nodes:
- cp_node3
- cp_node4

node1 have the following channels with inbound capacity amount:

- channel_1 for 100k sat:
node1 -> cp_node3

- channel_2 for 200k sat:
node1 -> cp_node3

node2 have the following channels with inbound capacity amount:

- channel_3 for 100k sat:
node2 -> cp_node3

- channel_4 for 50k sat:
node2 -> cp_node4

After the filtering is done we want the invoice to include the following channels in the hints (plus the RouteHintHop with src_node_id: hint.real_node_pubkey):
- channel_2
- channel_3
- channel_4

In other words, the highest inbound capacity channel for every counterparty node for each of our participating phantom payment-receiving nodes.

I previously thought we wanted to only include one channel per phantom payment-receiving node, which would have resulted in channel_2 and channel_3 (which is what the currently implemented code does).

1. Is this new filtering logic explained above correct?

2. Given the latest feedback, do I understand you correctly that if a phantom payment-receiving node has any private channel, we would like to just push a single RouteHintHop to the invoice for that node with the `src_node_id`: `hint.real_node_pubkey` and the `short_channel_id`: `hint.phantom_scid`?

Ps. if I'm completely incorrect about the filtering for phantom invoices are supposed to work, please let me know .

TheBlueMatt commented:
> In other words, the highest inbound capacity channel for every counterparty node for each of our participating phantom payment-receiving nodes.

Correct.

> Given the latest feedback, do I understand you correctly that if a phantom payment-receiving node has any private channel, we would like to just push a single RouteHintHop to the invoice for that node with the src_node_id: hint.real_node_pubkey and the short_channel_id: hint.phantom_scid, and let the sender find the path to the hint.real_node_pubkey phantom payment-receiving node by looking at the public channels?

Correct.

ViktorTigerstrom commented:
Just push some changes! If it's not suitable that commit a094d7d and commit e29c401 are part of this PR, let me know and I'll drop them and create a new PR with them and base this upon that PR.

TheBlueMatt commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -7586,33 +7586,7 @@ fn test_priv_forwarding_rejection() {
 
 	let chan_id_1 = create_announced_chan_between_nodes_with_value(&nodes, 0, 1, 1_000_000, 500_000_000, InitFeatures::known(), InitFeatures::known()).2;
 
-	// Note that the create_*_chan functions in utils requires announcement_signatures, which we do
```
Looks like you somehow ended up with some upstream changes here - can you rebase on upstream/main?

ViktorTigerstrom replied:
That specific line is actually intentionally changed by me, to use the new utility function I created in `lightning/src/ln/functional_test_utils.rs` in the functional_tests as well. Hence my comment above if it's suitable to include that commit e29c401 in this PR or if I should drop it :)

But I'll make sure to fetch the latest upstream and rebase this on that!

jkczyz commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -751,6 +751,38 @@ pub fn update_nodes_with_chan_announce<'a, 'b, 'c, 'd>(nodes: &'a Vec<Node<'b, '
 }
 
+pub fn create_private_chan_between_nodes_with_value<'a, 'b, 'c, 'd>(nodes: &'a Vec<Node<'b, 'c, 'd>>, a: usize, b: usize, channel_value: u64, push_msat: u64, a_flags: InitFeatures, b_flags: InitFeatures) -> (msgs::ChannelUpdate, msgs::ChannelUpdate, [u8; 32], Transaction) {
+	if nodes[a].node.get_current_default_configuration().channel_options.announced_channel {
+		panic!("`Nodes[a]` must have channel_options.announced_channel set to false")
+	}
+	nodes[a].node.create_channel(nodes[b].node.get_our_node_id(), channel_value, push_msat, 42, None).unwrap();
```
The last parameter to `create_channel` can be used to override the configuration values, I believe. So I think this function can simply override `channel_options.announced_channel` and re-use the other helpers instead of writing everything below this.

ViktorTigerstrom replied:
Thanks! Ok, I'll look into which other helpers I can use that doesn't require announcement signatures.

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -218,6 +198,85 @@ where
+		None => 0,
+	};
+	let route_hint_from_channel = |channel: ChannelDetails, short_channel_id: u64, forwarding_info: CounterpartyForwardingInfo| {
+        RouteHint(vec![RouteHintHop {
+            src_node_id: channel.counterparty.node_id,
+            short_channel_id,
+            fees: RoutingFees {
+                base_msat: forwarding_info.fee_base_msat,
+                proportional_millionths: forwarding_info.fee_proportional_millionths,
+            },
+            cltv_expiry_delta: forwarding_info.cltv_expiry_delta,
+            htlc_minimum_msat: None,
+            htlc_maximum_msat: None,}])
+    };
```
Please update to use tabs rather than spaces.

ViktorTigerstrom replied:
Oh my bad!

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -218,6 +198,85 @@ where
 	}
 }
 
+/// Utility to filter channels for an invoice, and return the corresponding route hints to include
```
nit: No need to say "Utility to". Simply say "Filters channels ... and returns ... ".

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -218,6 +198,85 @@ where
+/// Utility to filter channels for an invoice, and return the corresponding route hints to include
+/// in the invoice.
+///
+/// The filtering ensures that the RouteHints returned only include the highest inbound capacity
```
Use ticks on identifiers: `RouteHint`s. Likewise throughout.

ViktorTigerstrom replied:
Will do!

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -218,6 +198,85 @@ where
+/// in the invoice.
+///
+/// The filtering ensures that the RouteHints returned only include the highest inbound capacity
+/// channel per counterparty node for which channels are present. Channels with a lower inbound
```
Drop "for which channels are present" since the previous part of the sentence already establishes a channel exists.

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -218,6 +198,85 @@ where
+///
+/// Result:
+/// `Vec<RouteHint>`: The filtered RouteHints
+/// `bool`: True if any of the inputted channels are private, else false
```
s/inputted/given

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -218,6 +198,85 @@ where
+    };
+	for channel in channels {
+		if !channel.is_public {
+			// If any of the channels are private, return no hints and let the sender look at the
+			// public channels instead.
+			let has_private_channels = true;
+			return (vec![], has_private_channels);
```
If I understand the issue correctly, shouldn't we still return hints if *all* channels are private?

ViktorTigerstrom replied:
Hmmm that would make sense. I wasn't aware of that requirement, sorry. @TheBlueMatt, would you be able to confirm wether we would like that or not?

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -218,6 +198,85 @@ where
+/// `bool`: True if any of the inputted channels are private, else false
+fn filter_channels(channels: Vec<ChannelDetails>, min_inbound_capacity_required: Option<u64>) -> (Vec<RouteHint>, bool){
+	let mut filtered_channels: HashMap<PublicKey, (u64, RouteHint)> = HashMap::new();
+	let min_capacity_required = match min_inbound_capacity_required {
+		Some(amt) => amt,
+		None => 0,
+	};
```
The following is a little more idiomatic:

```rust
let min_inbound_capacity_required = min_inbound_capacity_required.unwrap_or(0);
```

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -218,6 +198,85 @@ where
+/// Result:
+/// `Vec<RouteHint>`: The filtered RouteHints
+/// `bool`: True if any of the inputted channels are private, else false
+fn filter_channels(channels: Vec<ChannelDetails>, min_inbound_capacity_required: Option<u64>) -> (Vec<RouteHint>, bool){
```
Would prefer to name this `min_inbound_capacity_msat`.

ViktorTigerstrom replied:
Ok makes sense!

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -218,6 +198,85 @@ where
+			let has_private_channels = true;
+			return (vec![], has_private_channels);
+		}
+		let short_channel_id = match channel.clone().short_channel_id {
```
Shouldn't need a clone for a primitive type.

ViktorTigerstrom replied:
Thanks! I'll make sure to do the referencing and cloning in a more suitable way.

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -218,6 +198,85 @@ where
+			Some(id) => id,
+			None => continue,
+		};
+		let forwarding_info = match channel.clone().counterparty.forwarding_info {
```
You can use references rather than cloning:

```rust
let forwarding_info = match &channel.counterparty.forwarding_info {
```

And then have the closure take references.

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -218,6 +198,85 @@ where
+				}
+				entry.insert((
+					channel.inbound_capacity_msat,
+					route_hint_from_channel(channel, short_channel_id, forwarding_info),
```
Pass `channel` by reference.

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -218,6 +198,85 @@ where
+			}
+		}
+	}
+	let route_hints = filtered_channels.iter()
+		.map(|(_channel, (_max_capacity, route_hint))| route_hint.clone())
+		.collect::<Vec<RouteHint>>();
```
You can avoid the clone by using `into_iter` instead of `iter`.

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -218,6 +198,85 @@ where
+			// If any of the channels are private, return no hints and let the sender look at the
+			// public channels instead.
+			let has_private_channels = true;
+			return (vec![], has_private_channels);
```
Rather than returning a `bool`, can the caller simply check if the returned `Vec` is empty?

ViktorTigerstrom replied:
That depends on on wether we would like the invoice to include single hop hint referencing a phantom payment-receiving node that only has public channels which inbound capacity are all below the invoice amt or not. 

Currently the code will not include a hint to those nodes. But If we're ok with including a hint to such nodes even though there are no channels that can be used to route to them (at least at the point of the invoice creation), that would certainly work!

That would definitely make the code cleaner :)

jkczyz replied:
Ah, that's an excellent point. I'd imagine we could have this method return `Result<Vec<RouteHint, ()>>` (or possible using a custom error type instead of `()`) and let the caller determine what to do if no channels have sufficient inbound capacity.

ViktorTigerstrom replied:
True! Ok I'll make the function return a `Result` type instead.

ViktorTigerstrom commented:
Thank a lot for the feedback @jkczyz! I'll make sure to update the PR asap. 

If possible, it would be great if I could get a clarification on:
* Wether we should include route hints if **all**  channels are private or not
* If we want the invoice to include a single hop route hint for a phantom payment-receiving node which only have channels with an inbound capacity lower than the invoice amt.

TheBlueMatt commented:
> If we should include route hints for a node with only private channels

Yes, that's the primary purpose of route hints :)

> If we want the invoice to include a single hop route hint for a phantom payment-receiving node which only has channels with inbound capacities lower than the invoice amt.

Yea, good question, at the risk of complexifying the PR, I think if *all* channels have less than the payment amount, we should just include all channels the same as we would if all channels had sufficient balance (still ignoring private channels if we have public ones).

ViktorTigerstrom commented:
Great, thanks for the clarifications! 

> Yea, good question, at the risk of complexifying the PR, I think if _all_ channels have less than the payment amount, we should just include all channels the same as we would if all channels had sufficient balance (still ignoring private channels if we have public ones).

Ok, I'll add that to the filtering function and also change the output to only return an empty `Vec` if we have a mix between public and private channels.

ViktorTigerstrom commented:
Updated the filtering logic per the feedback!

Note that I pushed the changes in separate commits to easier visualize to see the changes. Therefore the intermediate commits won't pass the tests/checks currently. But I'll squash them with the corresponding commit if they look good to fix that!

Please note that a phantom invoice will now include route hints for a phantom payment-receiving node with **NO** channels of higher inbound capacities than the invoice amt, even if another phantom payment-receiving node exists with such channels.

TheBlueMatt commented on lightning-invoice/src/utils.rs:
```diff
@@ -218,6 +189,95 @@ where
+	let mut filtered_channels: HashMap<PublicKey, (u64, RouteHint)> = HashMap::new();
+	let min_inbound_capacity = min_inbound_capacity_msat.unwrap_or(0);
+
+	let mut is_public_values: HashSet<bool> = HashSet::new();
```
There's never any reason to have a `HashSet<bool>`, the table is on the heap and the pointers and seeds are substantially larger than just storing two bools on the stack here.

ViktorTigerstrom replied:
Ok thanks, I'll replace it with bools instead. It was an attempt to remove some lines of code, but given your feedback I agree that thats just silly :)

TheBlueMatt commented on lightning-invoice/src/utils.rs:
```diff
@@ -218,6 +189,95 @@ where
+			}
+		}
+	}
+	if is_public_values.contains(&false) && is_public_values.contains(&true) {
```
I don't think we actually care about if there are private channels or not here - if there are *any* public channels, we don't need to return any channels at all.

ViktorTigerstrom replied:
Oh, I must have totally misunderstood in which scenario we want to filter the hints for the invoice then, I'm sorry about that! 

Just to clarify based on this feedback: For a node with _only_ public channels (_not just a mix_ of public and private channels), we don't want to include any route hints in the invoice then? Even if that node has multiple channels to the same counterparty node, which otherwise would make the hints filtering signal which of those channels is the highest capacity one/has enough capacity for the invoice amt.

TheBlueMatt replied:
If we have public channels, then everyone can see the channel values because we've announced which UTXOs they're associated with, so there's no need to hint anything in the invoices.

ViktorTigerstrom replied:
Thanks for clarifying! I'm still learning about lightning, so sorry if I'm a bit confused about when hints should be included.

I'll update the filtering so that no hints are returned if _any_ channel is public, and I'll also need to update a bunch of the tests to account for that.

jkczyz replied:
@TheBlueMatt Isn't `ChannelDetails:: inbound_capacity_msat` actually the channel liquidity balance, which is not publicly known?

TheBlueMatt replied:
> Isn't ChannelDetails:: inbound_capacity_msat actually the channel liquidity balance, which is not publicly known?

Yes, I'm not sure your point here, though? Are you suggesting we shouldn't sort based on it because it could reveal which channels of ours have relatively higher balance? I think that's okay, actually, we're not leaking all that much and probing could discover this a bit better, I'd think, anyway.

TheBlueMatt replied:
No, and I'm not sure how it got linked here, but because of #1278 - our counterparty will generally not actually care which SCID is present, they'll use whatever has enough balance. Its kinda nice to give them the "best" one, but its really not all *that* important. Only LDK hasn't done #1278, as far as I understand.

jkczyz replied:
> I'm wondering, and I think this is what @jkczyz's question is referring to as well: Why doesn't it make sense to include route hints when our node only has public channels, if we're using non public info (the inbound capacity) to generate the hints?

Yeah, that was my thinking.

> No, and I'm not sure how it got linked here, but because of #1278 - our counterparty will generally not actually care which SCID is present, they'll use whatever has enough balance. Its kinda nice to give them the "best" one, but its really not all _that_ important. Only LDK hasn't done #1278, as far as I understand.

Hmm... can't these channels be with different counterparties?

ViktorTigerstrom replied:
> Hmm... can't these channels be with different counterparties?

Yes, that's correct!
 
If a node has 2 public channels with 2 different counterparty nodes, and one of the channels inbound capacity is below the invoice amt, the current code will only include the channel to the other counterparty that's above the invoice amount in the hints.
The test `test_channels_with_lower_inbound_capacity_than_invoice_amt_hints_filtering`, shows that behavior in action :)

TheBlueMatt replied:
> Why doesn't it make sense to include route hints when our node only has public channels, if we're using non public info (the inbound capacity) to generate the hints?

Ah, I see your point. Yea, I mean that's not entirely unreasonable, but its also not crazy to just have them retry if they try to use a channel that doesn't have enough capacity. My other concern here is that if a node has some public channels there's a good chance they're a routing node and they have *lots* of public channels. One goal is to cut down the number of channels listed in the invoice, which can prevent QR code usage, and its easier to just include nothing if they're a routing node than it is to try to have some channel count limit behavior.

ViktorTigerstrom replied:
Ah, that makes sense! Ok I'll update the PR asap

TheBlueMatt reviewed (commented):
Looks basically good, I think.

ViktorTigerstrom commented:
Updated the code based on latest feedback, which makes the invoice filtering include no hints for a node if it has any public channels.

If it looks good, I'll squash the commits and also rebase on upstream as I see that there are now conflicts with upstream.

TheBlueMatt commented:
This needs rebase, sadly.

TheBlueMatt commented:
The bulk of the code LGTM, though, haven't reviewed the tests in detail.

TheBlueMatt commented on lightning-invoice/src/utils.rs:
```diff
@@ -218,6 +189,91 @@ where
+			htlc_maximum_msat: None,}])
+	};
+	for channel in channels {
+		if channel.is_public {
```
Oh, yea, I mean wont matter much, but might as well move this check to after the `forwarding_info` check/continue.

ViktorTigerstrom replied:
Ok great thanks! Fixed this in the latest pushed version

ViktorTigerstrom commented:
Rebased on upstream, and therefore also added scid alias support. 

In the latest upstream, I also found that a util to create unannounced channels had been added.  I therefore dropped my version, and used the upstream one for this PR.

Let me know if there's anything more that needs to be fixed :)

TheBlueMatt commented on lightning-invoice/src/utils.rs:
```diff
@@ -300,24 +356,28 @@ mod test {
 		let node_chanmgrs = create_node_chanmgrs(2, &node_cfgs, &[None, None]);
 		let nodes = create_network(2, &node_cfgs, &node_chanmgrs);
-		let _chan = create_announced_chan_between_nodes(&nodes, 0, 1, InitFeatures::known(), InitFeatures::known());
+		let mut private_chan_cfg = UserConfig::default();
```
Looks like this is now unused.

TheBlueMatt reviewed (approved):
These tests are great, thanks! Two trivial comments, LGTM!

ViktorTigerstrom commented:
Thanks! Fixed those comments :)

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -221,6 +201,91 @@ where
+/// If any public channel exists, the function returns no RouteHints, and the sender will need to
+/// look at the public channels to find a path instead.
+///
+/// Input:
+/// `channels`: The channels to filter.
+/// `min_inbound_capacity_msat`: Defines the lowest inbound capacity channels must have to not
+/// be filtered out, if any other channel above that amount exists.
+///
+/// Result:
+/// `Vec<RouteHint>`: The filtered `RouteHints`, which will be empty if any public channel exists.
```
I'd recommend removing this especially if it's repeating what's in the previous paragraphs in a slightly different way. Prefer to say it just once in a succinct way without loss of meaning.

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -437,6 +439,207 @@ mod test {
+		scid_aliases.insert(chan_1_0.0.short_channel_id_alias.unwrap());
+		scid_aliases.insert(chan_2_0.0.short_channel_id_alias.unwrap());
+
+		match_invoice_routes(
+			Some(5000),
+			&nodes[0],
+			scid_aliases,
+		);
```
Similarly, use one line here and elsewhere.

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -67,39 +67,30 @@ pub fn create_phantom_invoice<Signer: Sign, K: Deref>(
-					}])
-			);
+		let mut route_hints = filter_channels(
+			hint.channels.clone(),
```
Since the function takes hints by value, you can avoid this `clone` by using the de-structuring syntax in the for loop:

```rust
for PhantomRouteHints { channels, phantom_scid, real_node_pubkey } in phantom_route_hints {
```

ViktorTigerstrom replied:
Ah that's neat! Thanks for letting me know

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -437,6 +439,207 @@ mod test {
+		for hint in hints {
+			let hint_short_chan_id = (hint.0).0[0].short_channel_id;
+			assert!(chan_ids_to_match.contains(&hint_short_chan_id));
+			chan_ids_to_match.remove(&hint_short_chan_id);
```
Did you want to check that `chan_ids_to_match` is empty after iterating through the hints? This could be more useful than the earlier length check because you could include in the assertion failure which channels had been missed.

ViktorTigerstrom replied:
True!

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -437,6 +439,207 @@ mod test {
+		invoice_amt: Option<u64>,
+		invoice_node: &Node<'a, 'b, 'c>,
+		mut chan_ids_to_match: HashSet<u64>
+	){
```
nit: space between: `) {`

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -437,6 +439,207 @@ mod test {
+
+		for hint in hints {
+			let hint_short_chan_id = (hint.0).0[0].short_channel_id;
+			assert!(chan_ids_to_match.contains(&hint_short_chan_id));
+			chan_ids_to_match.remove(&hint_short_chan_id);
```
Could combine these by checking the result of `remove` instead:

```rust
`assert!(chan_ids_to_match.remove(&hint_short_chan_id))`
```

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -748,4 +748,333 @@ mod test {
+				},
+				2 => {
+					let hint_short_chan_id = hints[0].short_channel_id;
+					assert!(chan_ids_to_match.contains(&hint_short_chan_id));
+					chan_ids_to_match.remove(&hint_short_chan_id);
```
Similarly here.

jkczyz reviewed (commented):
Yes, looks much simpler and shaping up!

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -221,6 +201,91 @@ where
+/// in the invoice.
+///
+/// The filtering ensures that the `RouteHints` returned only include the highest inbound capacity
+/// channel per counterparty node. If any channel with a higher inbound capacity than the given
+/// `min_inbound_capacity_msat` exists, channels to other counterparty nodes with a lower inbound
+/// capacity than `min_inbound_capacity_msat` will be filtered out, even if they are the highest
+/// inbound capacity channel for that specific counterparty node.
```
Alternatively, for this comment and the one that follows, using the list from the commit message would suffice. I thought that was pretty well written, FWIW.

ViktorTigerstrom commented:
Thanks a lot for the feedback! Will try to address it shortly :)

ViktorTigerstrom commented on lightning-invoice/src/utils.rs:
```diff
@@ -161,30 +161,10 @@ where
-			htlc_maximum_msat: None,
-		}]));
-	}
+	let route_hints = filter_channels(
+		channelmanager.list_usable_channels(),
+		amt_msat,
+	);
```
Ok thanks for clarifying, will keep that in mind from now on

ViktorTigerstrom commented on lightning-invoice/src/utils.rs:
```diff
@@ -221,6 +201,91 @@ where
+				}
+				entry.insert((
+					channel.inbound_capacity_msat,
+					route_hint_from_channel(&channel, short_channel_id, forwarding_info),
```
Will do!

ViktorTigerstrom commented:
Pushed changes that addresses the feedback. I pushed the cleanup of the filtering function in a separate commit that I'll squash if it looks good.

ViktorTigerstrom commented on lightning-invoice/src/utils.rs:
```diff
@@ -221,6 +186,74 @@ where
+/// * If any public channel exists, the returned `RouteHint`s will be empty, and the sender will
+/// need to find the path by looking at the public channels instead
+fn filter_channels(channels: Vec<ChannelDetails>, min_inbound_capacity_msat: Option<u64>) -> Vec<RouteHint>{
+	let mut filtered_channels: HashMap<PublicKey, ChannelDetails> = HashMap::new();
```
I wasn't able to make this take a borrowed `&ChannelDetails`, as the borrow doesn't live outside of the channel loop context. Is that ok?

jkczyz replied:
Instead of borrowing `channel` directly within the loop, you can change the loop to iterate over references using:

```rust
for channel in channels.iter() {
```

Otherwise, `into_iter` is implicitly used, which consumes `channels` and yields values instead of references.

ViktorTigerstrom replied:
Oh nice, thanks for clarifying!

ViktorTigerstrom commented on lightning-invoice/src/utils.rs:
```diff
@@ -221,6 +186,74 @@ where
+	let mut min_capacity_channel_exists = false;
+
+	for channel in channels {
+		if channel.get_inbound_payment_scid().is_none() || channel.counterparty.forwarding_info.is_none() {
```
Made this over 100 characters intentionally as I thought it looked cleaner than the alternative.

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -417,6 +419,169 @@ mod test {
+			let hint_short_chan_id = (hint.0).0[0].short_channel_id;
+			assert!(chan_ids_to_match.remove(&hint_short_chan_id));
+		}
+		assert!(chan_ids_to_match.is_empty());
```
Could add a message so that unmatched ids are printed, something like this:

```rust
 assert!(chan_ids_to_match.is_empty(), "Unmatched short channel ids: {:?}", chan_ids_to_match);
```

ViktorTigerstrom replied:
Ah snap, missed that one.. thanks!

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -687,4 +687,332 @@ mod test {
+				_ => panic!("Incorrect hint length generated")
+			}
+		}
+		assert!(chan_ids_to_match.is_empty());
```
Likewise here.

jkczyz reviewed (commented):
Looking good. Thanks for bearing with me. Just some minor improvements left.

ViktorTigerstrom commented:
Fixed the latest feedback, and squashed the temporary commit
